src/api/wallet/subscriptionHelpers.ts

```diff
@@ -0,0 +1,138 @@
+import Web3 from 'web3'
+import { Command } from 'types'
+import { BlockHeader, Syncing } from 'web3-eth'
+import { LogsOptions, Log } from 'web3-core'
+import { Subscription } from 'web3-core-subscriptions'
+
+import { log } from 'utils'
+
+const DEFAULT_BLOCK_INTERVAL = 10000 //ms
+
+type SubscribeEvent = 'pendingTransactions' | 'syncing' | 'newBlockHeaders' | 'logs'
+
+interface SubscribeParams<T extends SubscribeEvent> {
+  web3: Web3
+  event: T
+  logOptions?: T extends 'logs' ? LogsOptions : void
+  callback: (data: Event2Data[T]) => void
+}
+
+interface Event2Data {
+  pendingTransactions: string
+  syncing: Syncing
+  newBlockHeaders: BlockHeader
+  logs: Log
+}
+
+// detects if providerr supports any given subscription
+const createWeb3Subscription = <T extends SubscribeEvent>({
+  web3,
+  event,
+  logOptions,
+  callback,
+}: SubscribeParams<T>): Promise<() => void> => {
+  return new Promise<Subscription<Event2Data[T]>>((resolve, reject) => {
+    const detectValidSubCb = (e: Error): void => {
+      if (e) {
+        reject(e)
+        return
+      }
+    }
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const params: [any, any] = (logOptions && event === 'logs'
```

Comment:
The `createWeb3Subscription` implementation is difficult to follow. Comments of intent would help.